### PR TITLE
using copy replace with directly object

### PR DESCRIPTION
**Pull request by zouyee**
**What this PR does / why we need it**:
We would not mutate the set during the sync if using the copy object. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```NONE

```

### CODE REVIEW

File: pkg/controller/statefulset/stateful_set.go [hunk 1]
```diff
@@ -205,6 +205,20 @@ func (ssc *StatefulSetController) updatePod(old, cur interface{}) {
 		return
 	}
 
+	if curPod.DeletionTimestamp != nil {
```
Comment (dixudx) [inline pkg/controller/statefulset/stateful_set.go]: Don't do this.
We've already got other function to handle these to-be-deleted pods.

File: pkg/controller/statefulset/stateful_set.go [hunk 2]
```diff
@@ -205,6 +205,20 @@ func (ssc *StatefulSetController) updatePod(old, cur interface{}) {
...
+		// on a restart of the controller manager, it's possible a new pod shows up in a state that
+		// is already pending deletion. Prevent the pod from being a creation observation.
+		ssc.deletePod(curPod)
+		return
+	}
+
+	if oldPod.DeletionTimestamp != nil {
+		// on a restart of the controller manager, it's possible a new pod shows up in a state that
```
Comment (dixudx) [inline pkg/controller/statefulset/stateful_set.go]: Same as above.

File: pkg/controller/statefulset/stateful_set.go [hunk 3]
```diff
@@ -422,6 +436,7 @@ func (ssc *StatefulSetController) sync(key string) error {
 		return err
 	}
 	set, err := ssc.setLister.StatefulSets(namespace).Get(name)
+	setCopy := set.DeepCopy()
```
Comment (dixudx) [inline pkg/controller/statefulset/stateful_set.go]: I can't see this set is mutated.

**Review by dixudx — CHANGES_REQUESTED** [review]

File: pkg/controller/statefulset/stateful_set.go (hunk 1 above)
Comment (yastij) [reply pkg/controller/statefulset/stateful_set.go]: +1 this shouldn't be done here

File: pkg/controller/statefulset/stateful_set.go (hunk 3 above)
Comment (yastij) [reply pkg/controller/statefulset/stateful_set.go]: The set is only used to adopt orphaned controller revisions and for pods listing + adoption by controllerRefManager. So we do not mutate the set here.
Also to sync the satetfulSet we currently deepCoy the set in `#syncStatefulSet()` due to some mutation of the set happening on the `#updateStatefulSet()` method (not sure it is still happening so cc @enisoc @smarterclayton)

**Review by yastij — CHANGES_REQUESTED** [review]

File: pkg/controller/statefulset/stateful_set.go [hunk 4]
```diff
@@ -205,6 +205,20 @@ func (ssc *StatefulSetController) updatePod(old, cur interface{}) {
 		return
 	}
 
+	if curPod.DeletionTimestamp != nil {
+		// on a restart of the controller manager, it's possible a new pod shows up in a state that
+		// is already pending deletion. Prevent the pod from being a creation observation.
+		ssc.deletePod(curPod)
```
Comment (janetkuo) [inline pkg/controller/statefulset/stateful_set.go]: We only need this in `addPod` handler (we already had this logic there).

**Review by kow3ns — CHANGES_REQUESTED** [review]
This PR needs a better explanation. Looking at the code,  I don't think that it correctly implements what the author believes it does.

**Pull request closed without merging**